runtime/lua/vim/filetype/detect.lua

```diff
@@ -1204,4 +1286,269 @@ end
+    local ft = type(v) == 'table' and v[1] or v
+    local opts = type(v) == 'table' and v[2] or {}
+    if opts.vim_regex and matchregex(name, k) or name:find(k) then
+      return ft
+    end
+  end
+end
+
+local patterns_text = {
+  ['^#compdef\\>'] = { 'zsh', { vim_regex = true } },
+  ['^#autoload\\>'] = { 'zsh', { vim_regex = true } },
+  -- ELM Mail files
+  ['^From [a-zA-Z][a-zA-Z_0-9%.=%-]*(@[^ ]*)? .* 19%d%d$'] = 'mail',
+  ['^From [a-zA-Z][a-zA-Z_0-9%.=%-]*(@[^ ]*)? .* 20%d%d$'] = 'mail',
+  ['^From %- .* 19%d%d$'] = 'mail',
+  ['^From %- .* 20%d%d$'] = 'mail',
+  -- Mason
+  ['^<[%%&].*>'] = 'mason',
+  -- Vim scripts (must have '" vim' as the first line to trigger this)
+  ['^" *[vV]im$['] = 'vim',
+  -- libcxx and libstdc++ standard library headers like ["iostream["] do not have
+  -- an extension, recognize the Emacs file mode.
+  ['%-%*%-.*[cC]%+%+.*%-%*%-'] = 'cpp',
+  ['^\\*\\* LambdaMOO Database, Format Version \\%([1-3]\\>\\)\\@!\\d\\+ \\*\\*$'] = {
+    'moo',
+    { vim_regex = true },
+  },
+  -- Diff file:
+  -- - "diff" in first line (context diff)
+  -- - "Only in " in first line
+  -- - "--- " in first line and "+++ " in second line (unified diff).
+  -- - "*** " in first line and "--- " in second line (context diff).
+  -- - "# It was generated by makepatch " in the second line (makepatch diff).
+  -- - "Index: <filename>" in the first line (CVS file)
+  -- - "=== ", line of "=", "---", "+++ " (SVK diff)
+  -- - "=== ", "--- ", "+++ " (bzr diff, common case)
+  -- - "=== (removed|added|renamed|modified)" (bzr diff, alternative)
+  -- - "# HG changeset patch" in first line (Mercurial export format)
+  ['^\\(diff\\>\\|Only in \\|\\d\\+\\(,\\d\\+\\)\\=[cda]\\d\\+\\>\\|# It was generated by makepatch \\|Index:\\s\\+\\f\\+\\r\\=$\\|===== \\f\\+ \\d\\+\\.\\d\\+ vs edited\\|==== //\\f\\+#\\d\\+\\|# HG changeset patch\\)'] = {
+    'diff',
+    { vim_regex = true },
+  },
+  function(contents)
+    return diff(contents)
+  end,
+  -- PostScript Files (must have %!PS as the first line, like a2ps output)
+  ['^%%![ \t]*PS'] = 'postscr',
+  function(contents)
+    return m4(contents)
+  end,
+  -- SiCAD scripts (must have procn or procd as the first line to trigger this)
+  ['^ *proc[nd] *$'] = { 'sicad', { ignore_case = true } },
+  ['^%*%*%*%*  Purify'] = 'purifylog',
+  -- XML
+  ['<%?%s*xml.*%?>'] = 'xml',
+  -- XHTML (e.g.: PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN")
+  ['\\<DTD\\s\\+XHTML\\s'] = 'xhtml',
+  -- HTML (e.g.: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN")
+  -- Avoid "doctype html", used by slim.
+  ['\\c<!DOCTYPE\\s\\+html\\>'] = { 'html', { vim_regex = true } },
+  -- PDF
+  ['^%%PDF%-'] = 'pdf',
+  -- XXD output
+  ['^%x%x%x%x%x%x%x: %x%x ?%x%x ?%x%x ?%x%x '] = 'xxd',
+  -- RCS/CVS log output
+  ['^RCS file:'] = { 'rcslog', { start_lnum = 1, end_lnum = 2 } },
+  -- CVS commit
+  ['^CVS:'] = { 'cvs', { start_lnum = 2 } },
+  ['^CVS: '] = { 'cvs', { start_lnum = -1 } },
```

Comment:
I don't think that works, though, since you use `start_lnum` to access the `contents_array` directly (rather than the `get_lines` function).